### PR TITLE
8327150: [lworld] Update C1 to support new value construction scheme from JEP 401

### PR DESCRIPTION
**Pull request by fparain**
Fix value creation in C1, remove the old special node for value creation and out everything under the NewInstance node.
There's also a fix in can_be_inline_klass() method that was causing a wrong behavior for acmp.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8327150](https://bugs.openjdk.org/browse/JDK-8327150): [lworld] Update C1 to support new value construction scheme from JEP 401 (**Bug** - P2)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1029/head:pull/1029` \
`$ git checkout pull/1029`

Update a local copy of the PR: \
`$ git checkout pull/1029` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1029/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1029`

View PR using the GUI difftool: \
`$ git pr show -t 1029`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1029.diff">https://git.openjdk.org/valhalla/pull/1029.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1029#issuecomment-1973912471)

### CODE REVIEW

File: src/hotspot/share/c1/c1_Runtime1.cpp [hunk 1]
```diff
@@ -369,24 +369,22 @@ static void allocate_instance(JavaThread* current, Klass* klass, TRAPS) {
   h->check_valid_for_instantiation(true, CHECK);
   // make sure klass is initialized
   h->initialize(CHECK);
-  // allocate instance and return via TLS
-  oop obj = h->allocate_instance(CHECK);
+  oop obj = nullptr;
+  if(h->is_empty_inline_type()) {
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/c1/c1_Runtime1.cpp:373]: Fixed.

File: src/hotspot/share/c1/c1_Runtime1.cpp [hunk 2]
```diff
@@ -369,24 +369,22 @@ static void allocate_instance(JavaThread* current, Klass* klass, TRAPS) {
...
+    assert(h->is_inline_klass(), "Sanity check");
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/c1/c1_Runtime1.cpp:374]: Removed.

File: src/hotspot/share/c1/c1_Runtime1.cpp [hunk 3]
```diff
@@ -369,24 +369,22 @@ static void allocate_instance(JavaThread* current, Klass* klass, TRAPS) {
...
+    assert(InlineKlass::cast(h)->default_value() != nullptr, "");
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/c1/c1_Runtime1.cpp:375]: Fixed.

File: src/hotspot/share/c1/c1_Instruction.hpp [hunk 4]
```diff
@@ -2090,8 +2063,9 @@ LEAF(If, BlockEnd)
     s->append(fsux);
     set_sux(s);
     if (!_substitutability_check) {
-      assert(x->as_NewInlineTypeInstance() == nullptr || y->type() == objectNull, "Sanity check");
-      assert(y->as_NewInlineTypeInstance() == nullptr || x->type() == objectNull, "Sanity check");
+      // FIXME: Removal of NewInlineTypeInstance instruction makes those assert hard to write
+      // assert(x->as_NewInlineTypeInstance() == nullptr || y->type() == objectNull, "Sanity check");
+      // assert(y->as_NewInlineTypeInstance() == nullptr || x->type() == objectNull, "Sanity check");
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/c1/c1_Instruction.hpp:2068]: OK

File: src/hotspot/share/c1/c1_Compiler.cpp [hunk 5]
```diff
@@ -255,6 +255,9 @@ void Compiler::compile_method(ciEnv* env, ciMethod* method, int entry_bci, bool
     // of Compilation to occur before we release the any
     // competing compiler thread
     ResourceMark rm;
+    if (PrintCFGToFile) {  // borrowing flag for JDK build
+      tty->print_cr("Starting C1 compilation of %s.%s", method->holder()->name()->as_utf8(), method->name()->as_utf8());
+    }
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/c1/c1_Compiler.cpp:260]: Oops, remaining of tracing during debugging. Removed.

**Comment by fparain** [comment]
Those fixes allow many C1 tests to pass, for instance all tests in TestC1.java now pass in scenarios #0 and #3, but the whole test fails because of scenarios #1 and #4 which involve C2.
Running all tests in test/hotspot/jtreg/compiler/valhalla/inlinetypes with `-XX:TieredStopAtLevel=3` shows many errors due to C2 (`assert(false) failed: Should have been buffered`), some verifier errors, and some tests not compliant with the new type hierarchy rules for value classes. But unless I missed them, I didn't notice failures directly caused by C1. However, it doesn't guarantee that once fixed, the problematic tests mentioned above will pass with C1.

**Review by TobiHartmann — APPROVED** [review]
Okay, thanks for the details. Looks good to me!

**Comment by fparain** [comment]
Thanks for the reviews Tobias.

**Pull request closed without merging**